ray_julia_jll/gen/artifacts.jl

```diff
@@ -0,0 +1,137 @@
+    -julia_version\+(?<julia_version>([0-9]\.){3})
+    tar.gz$
+    """x
+
+const GH_RELEASE_ASSET_PATH_REGEX = r"""
+    ^/(?<owner>[^/]+)/(?<repo_name>[^/]+)/
+    releases/download/
+    (?<tag>[^/]+)/(?<file_name>[^/]+)$
+    """x
+
+# Compute the Artifact.toml `git-tree-sha1`.
+function tree_hash_sha1(tarball_path)
+    return open(GzipDecompressorStream, tarball_path, "r") do tar
+        SHA1(Tar.tree_hash(tar))
+    end
+end
+
+# Compute the Artifact.toml `sha256` from the compressed archive.
+function sha256sum(tarball_path)
+    return open(tarball_path, "r") do tar
+        bytes2hex(sha256(tar))
+    end
+end
+
+function remote_url(repo_root::AbstractString, name::AbstractString="origin")
+    return LibGit2.with(LibGit2.GitRepo(repo_root)) do repo
+        LibGit2.with(LibGit2.lookup_remote(repo, name)) do remote
+            LibGit2.url(remote)
+        end
+    end
+end
+
+function get_release_asset_urls(pkg_url, jll_version)
+    # e.g. "git@github.com:beacon-biosignals/ray.jl"
+    _, pkg = split(pkg_url, ":")
+    assets_url = joinpath(GITHUB_URL,"$pkg", "releases", "tags", "v$(jll_version)")
+    io = IOBuffer()
+    run(pipeline(`$(curl()) $assets_url`, `$(jq()) -r '.assets[].browser_download_url'`, io))
+    assets = split(String(take!(io)), "\n"; keepempty=false)
+    return assets
+end
+
+function download_asset(asset_url)
+    run(`$(wget()) $asset_url -P $DIR`)
+    filename = replace(basename(asset_url), "%2B" => "+") # TODO: better way to parse URL in unicode?
+    return joinpath(DIR, filename)
+end
+
+if abspath(PROGRAM_FILE) == @__FILE__
+
+    repo_path = abspath(joinpath(@__DIR__, "..", ".."))
+    pkg_url = remote_url(repo_path)
+
+    jll_path = joinpath(repo_path, "ray_julia_jll")
+    artifacts_toml = joinpath(jll_path, "Artifacts.toml")
+
+    # Read Project.toml
+    jll_project_toml = joinpath(jll_path, "Project.toml")
+    jll_project = read_project(jll_project_toml)
+    jll_version = jll_project.version
+
+    wrappers_dir = joinpath(jll_path, "src", "wrappers")
+
+    artifacts_urls = get_release_asset_urls(jll_version)
+
+    for artifact_url in artifacts_urls
+
+        artifact_path = download_asset(artifact_url)
+
+        m = match(TRIPLET_REGEX, basename(artifact_path))
+        if isnothing(m)
+            throw(ArgumentError("Could not parse host triplet from $(basename(artifact_path))"))
+        end
+
+        platform_triplet = m[:triplet]
+        julia_version = m[:julia_version]
+        platform = parse(BinaryPlatforms.Platform, platform_triplet)
+
+        bind_artifact!(
+            artifacts_toml,
+            "ray_julia",
+            tree_hash_sha1(artifact_path);
+            platform=platform,
+            download_info=[(artifact_url, sha256sum(artifact_path))],
+            force=true
+        )
+
+        host_wrapper = joinpath(wrappers_dir, "$platform_triplet-julia_version+$julia_version.jl")
+        cp("wrapper.jl.tmp", host_wrapper; force=true)
```

Comment:
Probably would be better to move platform agnostic code out of the platform specific wrappers? You could probably just make a `src/wrappers/common.jl` which all the platform specific wrappers just include.

Reply:
good shout - didn't really look at how much boilerplate there was